[CI] DirectBlobContainerIndexInputTests.testRandomReads

**Build scan**:
https://gradle-enterprise.elastic.co/s/j4pqslngn7h5e

**Repro line**:
```
./gradlew ':x-pack:plugin:searchable-snapshots:test' --tests "org.elasticsearch.index.store.direct.DirectBlobContainerIndexInputTests.testRandomReads" \
  -Dtests.seed=ED84AFC0A8C9EBE1 \
  -Dtests.security.manager=true \
  -Dtests.locale=sr-RS \
  -Dtests.timezone=Europe/Volgograd \
  -Druntime.java=8
```

**Reproduces locally?**:
YES

**Applicable branches**:
7.x

**Failure history**:
1st occurence

**Failure excerpt**:
```
org.elasticsearch.index.store.direct.DirectBlobContainerIndexInputTests > testRandomReads FAILED |  
-- | --
  | arrays first differed at element [1131]; expected:<40> but was:<0> |  
  | at __randomizedtesting.SeedInfo.seed([ED84AFC0A8C9EBE1:D8860D3342654FCB]:0) |  
  | at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:55) |  
  | at org.junit.Assert.internalArrayEquals(Assert.java:532) |  
  | at org.junit.Assert.assertArrayEquals(Assert.java:341) |  
  | at org.junit.Assert.assertArrayEquals(Assert.java:352) |  
  | at org.elasticsearch.index.store.direct.DirectBlobContainerIndexInputTests.testRandomReads(DirectBlobContainerIndexInputTests.java:140) |  
  |   |  
  | Caused by: |  
  | java.lang.AssertionError: expected:<40> but was:<0> |  
  | at org.junit.Assert.fail(Assert.java:88) |  
  | at org.junit.Assert.failNotEquals(Assert.java:834) |  
  | at org.junit.Assert.assertEquals(Assert.java:118) |  
  | at org.junit.Assert.assertEquals(Assert.java:144) |  
  | at org.junit.internal.ExactComparisonCriteria.assertElementsEqual(ExactComparisonCriteria.java:8) |  
  | at org.junit.internal.ComparisonCriteria.arrayEquals(ComparisonCriteria.java:53) |  
  | ... 4 more
```

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/BaseSearchableSnapshotIndexInput.java
@@ -24,6 +24,7 @@ import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Predicate;
 
 import static org.elasticsearch.index.store.checksum.ChecksumBlobContainerIndexInput.checksumToBytesArray;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsUtils.toIntBytes;
 
 public abstract class BaseSearchableSnapshotIndexInput extends BufferedIndexInput {
 
@@ -99,7 +100,8 @@ public abstract class BaseSearchableSnapshotIndexInput extends BufferedIndexInpu
             return false;
         }
         final long position = getFilePointer() + this.offset;
-        if (position < fileInfo.length() - CodecUtil.footerLength()) {
+        final long checksumPosition = fileInfo.length() - CodecUtil.footerLength();
+        if (position < checksumPosition) {
             return false;
         }
         if (isClone) {
@@ -107,8 +109,12 @@ public abstract class BaseSearchableSnapshotIndexInput extends BufferedIndexInpu
         }
         boolean success = false;
         try {
+            final int checksumOffset = toIntBytes(Math.subtractExact(position, checksumPosition));
+            assert checksumOffset <= CodecUtil.footerLength() : checksumOffset;
+            assert 0 <= checksumOffset : checksumOffset;
+
             final byte[] checksum = checksumToBytesArray(fileInfo.checksum());
-            b.put(checksum, CodecUtil.footerLength() - remaining, remaining);
+            b.put(checksum, checksumOffset, remaining);
             success = true;
         } catch (NumberFormatException e) {
             // tests disable this optimisation by passing an invalid checksum

